### DISCUSSION TITLE
CDM function's parameter type cardinality not being checked?

### DISCUSSION THREAD

**Discussion by ghost** [finos/common-domain-model#3422] [Q&A] [answered]
## Support Question

I have come across couple of cases in CDM Rosetta function calls where the cardinality of the target function arguments appears to be ignored. 

For example, there are cases where a function declared to accept a value of type with cardinality [1..1], can be called with a constant [ ] or a value declared as [0..1]. Seems odd that when the function expects 1 and only 1 value, Rosetta allows it being called with no value.

Is that accepted for CDM/DRR? I'd imagine that could lead to some inconsistent behaviour.

Could someone please clarify?

Thanks

**Comment by lolabeis** [comment]
Hi, I will move this to a Discussion / Q&A and close this issue

**Comment by SimonCockx** [comment, marked as answer]
Hi @ghost, I know this is an old question, but just to come back to this. There is actually a good reason why these cardinality checks are loose.

Just to illustrate, consider the following function that takes a required argument.
```
func RequiredArgumentFunction:
  inputs:
    required string (1..1)
  output:
    result boolean (1..1)

  set result:
    required = "Dummy"
```

As you have noted correctly, this function can also be called with an optional value, e.g., given an optional attribute
```
optional string (0..1)
```
we can still call `RequiredArgumentFunction(optional)` without any errors. Suppose that we would forbid this, then consider another expression:
```
if optional exists
then RequiredArgumentFunction(optional)
```
This expression will only call the function if `optional` actually exists, so nothing can go wrong here. However, since we have forbidden calling the function with a cardinality constraint `(0..1)` Rosetta would show an error. This is not desired.

Of course, one could argue that Rosetta should be smarter and figure out if a check was performed, in this context the _actual_ constraint of `optional` is `(1..1)`, which should be valid. However, you soon run into more complicated situations where things are not that easy, e.g., when passing more complicated expressions.

This was one of the subjects of my [master thesis](https://github.com/SimonCockx/nouga/blob/83345efb35d50dab8674040e3aad4663c2c55208/thesis/final-digital.pdf) - the conclusion is that it it definitely possible theoretically, but quite an investment to get right. :)

**Comment by CDM-ReleaseManagement-LSG** [comment]
I hope the previous answer helped but I will close this for now.  Please open another Discussion or Issue, if you wish.
